### PR TITLE
*For Review Only* Apex 247 - Implementing Delay operator

### PR DESCRIPTION
**Pull request by sandeshh**
@davidyan74 Please review. I

### CODE REVIEW

File: library/src/main/java/com/datatorrent/lib/iteration/DelayOperator.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
+package com.datatorrent.lib.iteration;
```
Comment (tushargosavi) [inline library/src/main/java/com/datatorrent/lib/iteration/DelayOperator.java]: licence header

File: demos/iteration/src/test/java/com/datatorrent/demos/iteration/ApplicationTest.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+/**
+ *
+ */
+public class ApplicationTest
+{
+  @Test
+  public void testSomeMethod() throws Exception
+  {
+	  LocalMode lma = LocalMode.newInstance();
```
Comment (tushargosavi) [inline demos/iteration/src/test/java/com/datatorrent/demos/iteration/ApplicationTest.java:36]: Indentation is off

File: library/src/test/java/com/datatorrent/lib/iteration/DelayOperatorTest.java [hunk 3]
```diff
@@ -0,0 +1,174 @@
+package com.datatorrent.lib.iteration;
```
Comment (tushargosavi) [inline library/src/test/java/com/datatorrent/lib/iteration/DelayOperatorTest.java]: license header?

File: library/src/main/java/com/datatorrent/lib/iteration/DefaultDelayOperator.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+import java.io.IOException;
+import java.util.ArrayList;
+
+import com.datatorrent.api.Context;
+import com.datatorrent.api.DefaultInputPort;
+import com.datatorrent.api.DefaultOutputPort;
+import com.datatorrent.api.Operator;
+import com.datatorrent.lib.util.WindowDataManager;
+import com.datatorrent.netlet.util.DTThrowable;
+
```
Comment (chaithu14) [inline library/src/main/java/com/datatorrent/lib/iteration/DefaultDelayOperator.java:31]: Java Docs?

File: library/src/main/java/com/datatorrent/lib/iteration/DefaultDelayOperator.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+public class DefaultDelayOperator<T> implements Operator.DelayOperator, Operator.CheckpointListener
+{
```
Comment (chaithu14) [inline library/src/main/java/com/datatorrent/lib/iteration/DefaultDelayOperator.java:33]: Provide java docs for getter and setter methods.

File: library/src/main/java/com/datatorrent/lib/iteration/DefaultDelayOperator.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    return windowDataManager;
+  }
+
+  public void setWindowDataManager(WindowDataManager windowDataManager)
+  {
+    this.windowDataManager = windowDataManager;
+  }
+
+  private WindowDataManager windowDataManager;
+  private transient long currentWindowId;
```
Comment (chaithu14) [inline library/src/main/java/com/datatorrent/lib/iteration/DefaultDelayOperator.java:45]: Can you make the variables to protected, which of the variables doesn't have setter and getter ?

File: library/src/main/java/com/datatorrent/lib/iteration/DefaultDelayOperator.java [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+  public transient DefaultInputPort<T> input = new DefaultInputPort<T>() {
+    @Override
+    public void process(T t)
+    {
+      processTuple(t);
+    }
+  };
+
+  public transient DefaultOutputPort<T> output = new DefaultOutputPort();
+
```
Comment (chaithu14) [inline library/src/main/java/com/datatorrent/lib/iteration/DefaultDelayOperator.java:60]: Access specifier is missed?

File: library/src/main/java/com/datatorrent/lib/iteration/DefaultDelayOperator.java [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+  DefaultDelayOperator()
+  {
```
Comment (chaithu14) [inline library/src/main/java/com/datatorrent/lib/iteration/DefaultDelayOperator.java:62]: Assign default value to windowDataManager.

File: library/src/main/java/com/datatorrent/lib/iteration/DefaultDelayOperator.java [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+    ArrayList<T> recoveredData;
+    try {
+      recoveredData = (ArrayList<T>)this.windowDataManager.load(operatorContextId, windowId);
+      if (recoveredData == null) {
+        return;
+      }
+      for ( T tuple : recoveredData) {
+        processTuple(tuple);
+      }
+    } catch (IOException e) {
```
Comment (chaithu14) [inline library/src/main/java/com/datatorrent/lib/iteration/DefaultDelayOperator.java:97]: Why re-throwing the exception?

File: demos/iteration/src/test/resources/log4j.properties [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+log4j.appender.RFA.layout=org.apache.log4j.PatternLayout
+log4j.appender.RFA.layout.ConversionPattern=%d{ISO8601} [%t] %-5p %c{2} %M - %m%n
+log4j.appender.RFA.File=/tmp/app.log
+
+# to enable, add SYSLOG to rootLogger
+log4j.appender.SYSLOG=org.apache.log4j.net.SyslogAppender
+log4j.appender.SYSLOG.syslogHost=127.0.0.1
+log4j.appender.SYSLOG.layout=org.apache.log4j.PatternLayout
+log4j.appender.SYSLOG.layout.conversionPattern=${dt.cid} %-5p [%t] %c{2} %x - %m%n
+log4j.appender.SYSLOG.Facility=LOCAL1
```
Comment (gauravgopi123) [inline demos/iteration/src/test/resources/log4j.properties:36]: Do you need RFA and SYSLOG?

**Comment by gauravgopi123** [comment]
@sandeshh What are the next steps for this PR?

**Comment by sandeshh** [comment]
@gauravgopi123 This implementation is outdated because there is already default implementation in Core which doesn't need WindowDataManager. So updating it for N-DelayOperator. Do you want me to create a new pull request ?

**Comment by gauravgopi123** [comment]
Please close this

**Pull request closed without merging**